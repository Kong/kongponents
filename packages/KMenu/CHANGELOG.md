# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 6.3.0 (2021-11-23)


### Features

* **ktable/kcardcatalog:** make states slottable ([#438](https://github.com/Kong/kongponents/issues/438)) ([43eac93](https://github.com/Kong/kongponents/commit/43eac9337193fe47b28b27576d727e6cbbabad50))





# 6.2.0 (2021-11-23)


### Features

* **ktable/kcardcatalog:** emit cta click events ([#437](https://github.com/Kong/kongponents/issues/437)) ([29eb2a5](https://github.com/Kong/kongponents/commit/29eb2a53cf180935afee1ee10f1b88b462dc29ad))





## 6.1.19 (2021-11-23)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.18 (2021-11-23)


### Bug Fixes

* **ktable:** add missing style ([#435](https://github.com/Kong/kongponents/issues/435)) ([8782dee](https://github.com/Kong/kongponents/commit/8782dee82dafc68c1f956149a249b950359401b1))





## 6.1.17 (2021-11-22)


### Bug Fixes

* fix deps and build process ([#434](https://github.com/Kong/kongponents/issues/434)) ([9fdd180](https://github.com/Kong/kongponents/commit/9fdd180c9816bd4a43d4cd9229fc45b180ba5eed))





## 6.1.16 (2021-11-19)


### Bug Fixes

* **publish:** update node version in publish ([#433](https://github.com/Kong/kongponents/issues/433)) ([26cfe34](https://github.com/Kong/kongponents/commit/26cfe34b6fd9fbb3d61ffdb95412804bac88ba36))





## 6.1.15 (2021-11-19)


### Bug Fixes

* **build:** update lerna ignore flag ([#431](https://github.com/Kong/kongponents/issues/431)) ([768c5e4](https://github.com/Kong/kongponents/commit/768c5e4e78bfca118fe7b2d022778854e076b323))





## 6.1.14 (2021-11-19)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.13 (2021-11-19)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.12 (2021-11-19)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.11 (2021-11-19)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.10 (2021-11-18)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.9 (2021-11-18)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.8 (2021-11-18)


### Bug Fixes

* **css:** box-shadow does not render on ios ([#420](https://github.com/Kong/kongponents/issues/420)) ([c0ba7cb](https://github.com/Kong/kongponents/commit/c0ba7cb90bf3c158a1baa6b879d33230359db830))





## 6.1.7 (2021-11-17)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.6 (2021-11-17)


### Bug Fixes

* **ktextarea:** limit exceeded event ([#418](https://github.com/Kong/kongponents/issues/418)) ([f7cf421](https://github.com/Kong/kongponents/commit/f7cf421a4ded7f5b99eaa5a18d02a27d450020b7))





## 6.1.5 (2021-11-17)

**Note:** Version bump only for package @kongponents/kmenu





## 6.1.4 (2021-11-17)


### Bug Fixes

* general fixes ([#415](https://github.com/Kong/kongponents/issues/415)) ([5b4bb48](https://github.com/Kong/kongponents/commit/5b4bb484d74e75779d39cf901b2cdaa23818e8e7))





## 6.1.3 (2021-11-17)

**Note:** Version bump only for package @kongponents/kmenu
