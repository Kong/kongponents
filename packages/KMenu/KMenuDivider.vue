<template>
  <hr class="k-menu-item-divider">
</template>

<script>
export default {
  name: 'KMenuDivider'
}
</script>

<style lang="scss" scoped>
.k-menu-item-divider{
  position: relative;
  left: -24px;
  width: calc(100% + 24px);
  margin: 0;
}

</style>
