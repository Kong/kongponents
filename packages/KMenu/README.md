# @kongponents/kmenu

[![](https://img.shields.io/npm/v/@kongponents/kmenu.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kmenu)

```html
<KMenu :description="'hello world'">
  Hello from a slot
</KMenu>
```
