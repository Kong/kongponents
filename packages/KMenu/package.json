{
  "name": "@kongponents/kmenu",
  "version": "6.21.0",
  "description": "<Menu component>",
  "main": "dist/KMenu.umd.min.js",
  "componentName": "KMenu",
  "source": "index.js",
  "files": [
    "dist",
    "*.vue",
    "index.js"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^6.21.0",
    "@kongponents/kicon": "^6.21.0",
    "@kongponents/styles": "^6.21.0",
    "vue-uuid": "^2.0.2"
  }
}
