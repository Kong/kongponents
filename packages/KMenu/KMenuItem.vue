<template>
  <li
    :class="{'expand-item':!item.expandable}"
    :key="item.key"
    :data-testid="`k-menu-item-${item.value}`"
  >
    <div
      class="menu-title"
      @click="isOpen = !isOpen">
      <span
        :class="isOpen && item.expandable ? 'title-dark' : ''"
        class="span-menu-title">
        <slot name="menuTitle"> {{ item.title }}
        </slot>
      </span>
      <span

        class="span-icon-container">
        <KIcon
          v-if="item.expandable"
          :icon="isOpen ? 'chevronUp' : 'chevronDown'"
          color="var(--grey-400)"
          size="16"
        />
      </span>
      <slot/>
    </div>
    <div
      :class="isOpen ? 'd-block' : 'd-none'"
      class="menu-content">
      <slot
        v-if="item.itemType === 'String'"
        name="menuContent"> {{ item.itemType }}
      </slot>
      <slot
        v-else
        name="menuContent"> {{ item.itemType }}
      </slot>
    </div>
    <slot
      v-if="!menuDivider"
      name="menuDivider"><KMenuDivider/>
    </slot>
  </li>
</template>

<script>
import KMenuDivider from './KMenuDivider.vue'
import KIcon from '@kongponents/kicon/KIcon.vue'

export default {
  name: 'KMenuItem',
  components: { KMenuDivider, KIcon },
  props: {
    item: {
      type: Object,
      default: null
    },
    menuDivider: {
      type: Boolean,
      default: false
    }
  },

  data: () => {
    return {
      isOpen: false
    }
  }
}
</script>

<style scoped lang="scss">
li {
  margin: 0;
  cursor: pointer;
  white-space: nowrap;
  font-family: var(--font-family-sans);
  font-style: normal;
  font-weight: 400;
  font-size: 13px;
  line-height: 24px;
  color: var(--grey-500);
  position: relative;

  &:not(:last-child):after {
    content: " ";
    position: absolute;
    left: 50%;
    transform: translateX(-56%);
    bottom: 0;
    width: 90%;
    height: 1px;
    background-color: var(--grey-300);
  }
}

.menu-title {
  width: 100%;
  display: flex;
  padding: 8px 0;
  text-align: center;

  &:hover {
    color: var(--grey-600);
  }
}

.span-icon-container {
  margin-left: auto;
  margin-top: auto;
  margin-bottom: auto;
  padding-right: 23px;
  height: 24px;
  width: 24px;
}

.title-dark {
  color: var(--grey-600);
}

li.expand-item {
  cursor: default;
}
</style>
