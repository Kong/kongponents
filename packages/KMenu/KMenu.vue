<template>
  <div
    :style="widthStyle"
    class="k-menu">
    <slot name="body">
      <div>
        <template v-for="(item,index) in items">
          <KMenuItem
            :item="item"
            :expandable="item.expandable"
            :key="item.key"
            :type="item.type"
            :test-mode="testMode"
            :class="{'last-menu-item': index===items.length-1 }"
          />
        </template>
      </div>
    </slot>
    <div
      v-if="hasActionButton"
      class="clear-cta-button">
      <slot
        name="actionButton">
        <KButton
          @click="proceed"
          @keyup.enter="proceed"/>
      </slot>
    </div>
  </div>
</template>

<script>
import KMenuItem from './KMenuItem.vue'

export default {
  name: 'KMenu',
  components: { KMenuItem },
  props: {
    items: {
      type: Array,
      required: false,
      default: () => []
    },
    actionButton: {
      type: String,
      default: ''
    },
    width: {
      type: String,
      default: '284'
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    widthStyle: function () {
      return {
        width: this.width === 'auto' ? this.width : this.width + 'px'
      }
    },

    hasActionButton () {
      return !!this.$slots.actionButton
    }
  },

  methods: {
    proceed () {
      this.$emit('proceed')
    }
  }
}
</script>

<style scoped lang="scss">
.k-menu {
  background-color: var(--white);
  border: 1px solid var(--grey-300);
  border-radius: 4px;
  padding-top: 11px;
}

.clear-cta-button > button.k-button {
  border: none;
  color: var(--blue-300);
  font-weight: 500;
  font-size: 13px;
  line-height: 13px;
  margin-bottom: 6px;
  margin-top: 10px;
  padding-top: 2px;
  &:active, &:hover {
    color: var(--blue-500);
    background-color: transparent;
  }
  &:focus {
    box-shadow: none;
    color: var(--blue-500);
    background-color: transparent;
  }
}

</style>
