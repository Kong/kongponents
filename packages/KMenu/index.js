import KMenu from './KMenu'
import KMenuDivider from './KMenuDivider'
import KMenuItem from './KMenuItem'

const Plugin = {
  install (Vue) {
    Vue.component('KMenu', KMenu)
    Vue.component('KMenuDivider', KMenuDivider)
    Vue.component('KMenuItem', KMenuItem)
  }
}

export default Plugin
export { KMenu, KMenuDivider, KMenuItem }

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(Plugin)
}
