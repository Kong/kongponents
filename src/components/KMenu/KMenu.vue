<template>
  <div
    :style="widthStyle"
    class="k-menu"
  >
    <slot name="body">
      <div>
        <template
          v-for="(item, index) in items"
          :key="item.title + index"
        >
          <KMenuItem
            :item="item"
            :expandable="item.expandable"
            :type="item.type"
            :last-menu-item="index === items.length - 1"
            :test-mode="testMode"
            :class="{ 'last-menu-item': index === items.length - 1 }"
          />
        </template>
      </div>
    </slot>
    <div
      v-if="hasActionButton"
      class="clear-cta-button"
    >
      <KMenuDivider />
      <slot name="actionButton">
        <KButton
          @click="proceed"
          @keyup.enter="proceed"
        />
      </slot>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, PropType } from 'vue'
import KMenuItem from '@/components/KMenu/KMenuItem.vue'
import KMenuDivider from '@/components/KMenu/KMenuDivider.vue'

export interface KMenuItemType {
  title: string
  description?: string
  expandable: boolean
  type: 'string' | 'number' | 'divider'
}

export default defineComponent({
  name: 'KMenu',
  components: { KMenuItem, KMenuDivider },
  props: {
    items: {
      type: Array as PropType<KMenuItemType[]>,
      required: false,
      default: () => [],
    },
    actionButton: {
      type: String,
      default: '',
    },
    width: {
      type: String,
      default: '284',
    },
    /**
     * Test mode - for testing only, strips out generated ids
     */
    testMode: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['proceed'],
  setup(props, { emit, slots }) {
    const widthStyle = computed((): Record<string, string> => {
      return {
        width: props.width === 'auto' || props.width.includes('px') ? props.width : props.width + 'px',
      }
    })

    const hasActionButton = computed((): boolean => !!slots.actionButton)

    const proceed = (): void => {
      emit('proceed')
    }

    return {
      widthStyle,
      hasActionButton,
      proceed,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';

.k-menu {
  background-color: var(--white);
  border: 1px solid var(--grey-300);
  border-radius: 4px;
  padding-top: 11px;
  padding-bottom: 8px;
}

.clear-cta-button > :deep(button.k-button) {
  border: none;
  color: var(--blue-300);
  font-weight: 500;
  font-size: 13px;
  line-height: 13px;
  margin-bottom: 6px;
  margin-top: 10px;
  padding-top: 2px;

  &:active, &:hover {
    color: var(--blue-500);
    background-color: transparent;
  }

  &:focus {
    box-shadow: none;
    color: var(--blue-500);
    background-color: transparent;
  }
}

</style>
