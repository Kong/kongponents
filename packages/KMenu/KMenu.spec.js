import { mount } from '@vue/test-utils'
import KMenu from '@/KMenu/KMenu'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * propsData: {
 *   testMode: true
 * }
 */
describe('KMenu', () => {
  function getItems (count) {
    let myItems = []

    for (let i = 0; i < count; i++) {
      myItems.push({
        title: 'Item ' + i,
        type: Math.random() < 0.5 ? 'string' : 'number',
        expandable: Math.random() < 0.5,
        description: "The item's description for number " + i
      })
    }

    return myItems
  }

  const customItem = {
    title: 'Item #',
    description: 'Cras aliquet auctor ex ut hendrerit. Donec sagittis est nec aliquet semper. Quisque feugiat metus orci, at ullamcorper odio molestie non. Nam dignissim sed ligula ut commodo.'
  }

  describe('general', () => {
    it('renders proper menu when using props', () => {
      const wrapper = mount(KMenu, {
        propsData: {
          items: getItems(5),
          testMode: true
        }
      })

      expect(wrapper.find('.k-menu').exists()).toBeTruthy()
      expect(wrapper.findAll('.k-menu-item')).toHaveLength(5)
    })

    it('shows chevron down icon', () => {
      const wrapper = mount(KMenu, {
        propsData: {
          items: [customItem],
          expandable: true,
          testMode: true
        }
      })

      expect(wrapper.find('.k-menu').exists()).toBeTruthy()
      expect(wrapper.find('.k-menu-item .k-button > .span-icon-container').exists()).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
