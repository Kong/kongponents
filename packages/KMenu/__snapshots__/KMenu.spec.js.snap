// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KMenu general shows chevron down icon 1`] = `
<div class="k-menu" style="width: 284px;" expandable="true">
  <div>
    <div id="test-menuitem-id-1234" data-testid="Item-#-menu-item" test-mode="true" class="k-menu-item last-menu-item"><button type="button" class="k-button menu-button non-visual-button medium rounded outline" aria-labelledby="menuItemId"><span class="span-menu-title">
        Item #
      </span> <span class="span-icon-container"><!----></span>
        <!----></button>
      <div class="menu-content d-none">
        <!---->
      </div>
      <!---->
    </div>
  </div>
  <!---->
</div>
`;
